src/toil/wdl/wdltoil.py

```diff
@@ -1318,7 +1318,11 @@ def _virtualize_filename(self, filename: str) -> str:
         else:
             # Otherwise this is a local file and we want to fake it as a Toil file store file
             # Make it an absolute path
-            if self.execution_dir is not None:
+            parsed = urlparse(unquote(filename))
+            if parsed.scheme == "file":
+                # conversion was already done by normalize_uri
+                abs_filename = parsed.path
+            elif self.execution_dir is not None:
```

Comment:
No, this has to be in the *other* order: the encoding is applied to the path portion of the URL *before* it is assembled into the URL, so you should parse apart the URL and *then* decode the path. Otherwise characters in the path can sneak out and break the parsing of the URL. Like this:
```
>>> from urllib.parse import quote, unquote, urlparse
>>> url = "file://" + quote("/tmp/data/file.maybegood?.attempt.#1.dat")
>>> print(url)
file:///tmp/data/file.maybegood%3F.attempt.%231.dat
>>> print(urlparse(unquote(url)))
ParseResult(scheme='file', netloc='', path='/tmp/data/file.maybegood', params='', query='.attempt.', fragment='1.dat')
>>> print(urlparse(unquote(url)).path)
/tmp/data/file.maybegood
>>> print(urlparse(url))
ParseResult(scheme='file', netloc='', path='/tmp/data/file.maybegood%3F.attempt.%231.dat', params='', query='', fragment='')
>>> print(unquote(urlparse(url).path))
/tmp/data/file.maybegood?.attempt.#1.dat
```

So it should be like:

```suggestion
            parsed = urlparse(filename)
            if parsed.scheme == "file":
                # conversion was already done by normalize_uri
                abs_filename = unquote(parsed.path)
            elif self.execution_dir is not None:
```